typing/ctype.ml

```diff
@@ -1305,73 +1300,71 @@ let instance_class params cty =
+     that have no free univars.
+   * In this first pass, [visited] is used as a mapping for previously visited
+     nodes, and must already contain all the free univars in [ty].
+   * The remaining (univar-closed) parts of the type are then instantiated
+     with [copy] using a common [copy_scope].
+   The reason to work in two passes lies in recursive types such as:
+     [let h (x : < m : 'a. < n : 'a; p : 'b > > as 'b) = x#m]
+   The type of [x#m] should be:
+     [ < n : 'c; p : < m : 'a. < n : 'a; p : 'b > > as 'b > ]
+   I.e., the universal type variable ['a] is both instantiated as a fresh
+   type variable ['c] when outside of its binder, and kept as universal
+   when under its binder.
+   Assumption: in the first call to [copy_sep], all the free univars should
+   be bound by the same [Tpoly] node. This guarantees that they are only
+   bound when under this [Tpoly] node, which has no free univars, and as
+   such is not part of the separate copy. In turn, this allows the separate
+   copy to keep the sharing of the original type without breaking its
+   binding structure.
+ *)
+let rec copy_sep ~copy_scope ~fixed ~free ~may_share
+    ~(visited : type_expr TypeHash.t) (ty : type_expr) =
   let univars = free ty in
   if is_Tvar ty || may_share && TypeSet.is_empty univars then
     if get_level ty <> generic_level then ty else
     let t = newstub ~scope:(get_scope ty) in
-    delayed_copy :=
-      lazy (Transient_expr.set_stub_desc t (Tlink (copy copy_scope ty)))
-      :: !delayed_copy;
+    add_delayed_copy t copy_scope ty;
     t
   else try
-    let t, bound_t = List.assq (get_id ty) visited in
-    let dl = if is_Tunivar ty then [] else diff_list bound bound_t in
-    if dl <> [] && conflicts univars dl then raise Not_found;
-    t
+    TypeHash.find visited ty
   with Not_found -> begin
     let t = newstub ~scope:(get_scope ty) in
-    let desc = get_desc ty in
-    let visited =
-      match desc with
-        Tarrow _ | Ttuple _ | Tvariant _ | Tconstr _ | Tobject _ | Tpackage _ ->
-          (get_id ty, (t, bound)) :: visited
-      | Tvar _ | Tfield _ | Tnil | Tpoly _ | Tunivar _ ->
-          visited
-      | Tlink _ | Tsubst _ ->
-          assert false
-    in
-    let copy_rec = copy_sep ~copy_scope ~fixed ~free ~bound visited in
+    TypeHash.add visited ty t;
+    let copy_rec = copy_sep ~copy_scope ~fixed ~free ~visited in
     let desc' =
-      match desc with
+      match get_desc ty with
       | Tvariant row ->
           let more = row_more row in
           (* We shall really check the level on the row variable *)
           let keep = is_Tvar more && get_level more <> generic_level in
+          (* In that case we should keep the original, but we still
+             call copy to correct the levels *)
+          if keep then (add_delayed_copy t copy_scope ty; Tvar None) else
           let more' = copy_rec ~may_share:false more in
           let fixed' = fixed && (is_Tvar more || is_Tunivar more) in
```

Comment:
In the `copy` version there is one case related to GADTs where the `fixed` field of the copied row is in fact changed:

https://github.com/ocaml/ocaml/blob/88c3fc65774ec669d9ed8e6ad3ef136ba0594a40/typing/ctype.ml#L1089-L1095

do we need similar logic here?

Reply:
As indicated, the reason is explained in #6163.
Namely, if `row_more`is a `Tconstr`, the row should also be fixed. Apparently this is not properly enforced by reification. This should probably rather be done inside reification than inside `copy` but this is clearly orthogonal to this PR.

Reply:
My question was whether the same issue could occur within the input of `copy_sep`, instead of `copy`, as so we would need to patch fixedness in the same way for correctness. It is not clear to me that `copy_sep` will never encounter a Tconstr row variable with incorrect fixedness, but if it is to you then maybe that's good enough.

Reply:
As said above, if there is a need to fix that, this should not be in `copy_sep` but in the reification code.
Should try to come up with some example.

Reply:
My understanding of this point:

- currently this case in `copy` fixes a bug, but you believe the fix is not appropriate and should be done somewhere else (I agree with this intuition)
- by not reproducing the fix in `copy_sep`, we expose ourselves to a similar bug
- but you prefer to not duplicate the non-appropriate fix, and wait for later to do a better fix that would require no code in `copy` or `copy_sep`

That's okay with me, with the understanding that we decided to leave this bug open and fix it later.

Reply:
Yes, this is my view. With the recurring problem that, in order to fix a bug properly, you need a counter-example, and in this case it is not so easy to come up with `copy_sep`. Fixing reification should be easier to test.